Provide utilities to satisfy `LessThan` and `NumericRange`

### Describe the problem

So there's these new fun and interesting types we have to satisfy to use `error(status, body)`:

```ts
/**
 * @template {number} TNumber
 * @template {any[]} [TArray=[]]
 * @typedef {TNumber extends TArray['length'] ? TArray[number] : LessThan<TNumber, [...TArray, TArray['length']]>} LessThan
 */

/**
 * @template {number} TStart
 * @template {number} TEnd
 * @typedef {Exclude<TEnd | LessThan<TEnd>, LessThan<TStart>>} NumericRange
 */
```

However, TypeScript wasn't exactly made with these use cases in mind. For example, doing `if (x < 5)` doesn't change the type of `x` from `number` to `number < 5`.

Normally, if I wanted to implement a type guard to satisfy these new types, it could look something like this:

```ts
import type { LessThan, NumericRange } from '@sveltejs/kit'

export function isLessThan<Max extends number>(input: number, max: Max): input is LessThan<Max> {
	return input < max
}

export function toLessThan<Max extends number>(input: number, max: Max): LessThan<Max> {
	if (isLessThan(input, max)) return input
	return (max - 1) as LessThan<Max>
}

export function isNumericRange<Min extends number, Max extends number>(
	input: number,
	min: Min,
	max: Max,
): input is NumericRange<Min, Max> {
	return min <= input && input <= max
}

export function toNumericRange<Min extends number, Max extends number>(
	input: number,
	min: Min,
	max: Max,
): NumericRange<Min, Max> {
	if (isLessThan(input, min)) return min as NumericRange<Min, Max>
	if (isLessThan(input, max)) return input as NumericRange<Min, Max>
	return max as NumericRange<Min, Max>
}
```

However, this is exceedingly hard to achieve if you have a strict codebase.

```ts
/home/nato/Code/github.com/NatoBoram/Leanish/src/lib/utils/adapters.ts:3:74
Error: Type instantiation is excessively deep and possibly infinite. 

export function isLessThan<Max extends number>(input: number, max: Max): input is LessThan<Max> {
        return input < max


/home/nato/Code/github.com/NatoBoram/Leanish/src/lib/utils/adapters.ts:8:37
Error: Type instantiation is excessively deep and possibly infinite. 
export function toLessThan<Max extends number>(input: number, max: Max): LessThan<Max> {
        if (isLessThan(input, max)) return input
        return (max - 1) as LessThan<Max>


====================================
svelte-check found 2 errors and 0 warnings in 1 file
```
```ts
/home/nato/Code/github.com/NatoBoram/Leanish/src/lib/utils/adapters.ts
   8:30  error  Unsafe return of type `any` from function with return type `LessThan<Max>`           @typescript-eslint/no-unsafe-return
   9:2   error  Unsafe return of type `any` from function with return type `LessThan<Max>`           @typescript-eslint/no-unsafe-return
  25:30  error  Unsafe return of type `any` from function with return type `NumericRange<Min, Max>`  @typescript-eslint/no-unsafe-return
  26:30  error  Unsafe return of type `any` from function with return type `NumericRange<Min, Max>`  @typescript-eslint/no-unsafe-return
  27:2   error  Unsafe return of type `any` from function with return type `NumericRange<Min, Max>`  @typescript-eslint/no-unsafe-return

✖ 5 problems (5 errors, 0 warnings)
```
The amount of rules you have to break to achieve this is impractical.

### Describe the proposed solution

If these types are to stay, then it would be very useful to *also* provide type guards and adapters to easily satisfy them.

* `isLessThan`
* `toLessThan`
* `toNumericRange`
* `isNumericRange`

### Alternatives considered


### Importance

would make my life easier

### Additional Information

--- a/.changeset/mighty-spies-explain.md
+++ b/.changeset/mighty-spies-explain.md
@@ -1,0 +1,5 @@
+---
+"@sveltejs/kit": patch
+---
+
+fix: relax status code types

--- a/packages/kit/src/exports/index.js
+++ b/packages/kit/src/exports/index.js
@@ -3,6 +3,7 @@ import { BROWSER, DEV } from 'esm-env';
 
 export { VERSION } from '../version.js';
 
+// TODO 3.0: remove these types as they are not used anymore (we can't remove them yet because that would be a breaking change)
 /**
  * @template {number} TNumber
  * @template {any[]} [TArray=[]]
@@ -15,6 +16,9 @@ export { VERSION } from '../version.js';
  * @typedef {Exclude<TEnd | LessThan<TEnd>, LessThan<TStart>>} NumericRange
  */
 
+// Keep the status codes as `number` because restricting to certain numbers makes it unnecessarily hard to use compared to the benefits
+// (we have runtime errors already to check for invalid codes). Also see https://github.com/sveltejs/kit/issues/11780
+
 // we have to repeat the JSDoc because the display for function overloads is broken
 // see https://github.com/microsoft/TypeScript/issues/55056
 
@@ -23,10 +27,10 @@ export { VERSION } from '../version.js';
  * When called during request handling, this will cause SvelteKit to
  * return an error response without invoking `handleError`.
  * Make sure you're not catching the thrown error, which would prevent SvelteKit from handling it.
- * @param {NumericRange<400, 599>} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
  * @param {App.Error} body An object that conforms to the App.Error type. If a string is passed, it will be used as the message property.
  * @overload
- * @param {NumericRange<400, 599>} status
+ * @param {number} status
  * @param {App.Error} body
  * @return {never}
  * @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
@@ -37,10 +41,10 @@ export { VERSION } from '../version.js';
  * When called during request handling, this will cause SvelteKit to
  * return an error response without invoking `handleError`.
  * Make sure you're not catching the thrown error, which would prevent SvelteKit from handling it.
- * @param {NumericRange<400, 599>} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
  * @param {{ message: string } extends App.Error ? App.Error | string | undefined : never} [body] An object that conforms to the App.Error type. If a string is passed, it will be used as the message property.
  * @overload
- * @param {NumericRange<400, 599>} status
+ * @param {number} status
  * @param {{ message: string } extends App.Error ? App.Error | string | undefined : never} [body]
  * @return {never}
  * @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
@@ -51,7 +55,7 @@ export { VERSION } from '../version.js';
  * When called during request handling, this will cause SvelteKit to
  * return an error response without invoking `handleError`.
  * Make sure you're not catching the thrown error, which would prevent SvelteKit from handling it.
- * @param {NumericRange<400, 599>} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
  * @param {{ message: string } extends App.Error ? App.Error | string | undefined : never} body An object that conforms to the App.Error type. If a string is passed, it will be used as the message property.
  * @return {never}
  * @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
@@ -80,7 +84,7 @@ export function isHttpError(e, status) {
 /**
  * Redirect a request. When called during request handling, SvelteKit will return a redirect response.
  * Make sure you're not catching the thrown redirect, which would prevent SvelteKit from handling it.
- * @param {NumericRange<300, 308>} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages). Must be in the range 300-308.
+ * @param {300 | 301 | 302 | 303 | 304 | 305 | 306 | 307 | 308 | ({} & number)} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages). Must be in the range 300-308.
  * @param {string | URL} location The location to redirect to.
  * @throws {Redirect} This error instructs SvelteKit to redirect to the specified location.
  * @throws {Error} If the provided status is invalid.
@@ -91,7 +95,11 @@ export function redirect(status, location) {
 		throw new Error('Invalid status code');
 	}
 
-	throw new Redirect(status, location.toString());
+	throw new Redirect(
+		// @ts-ignore
+		status,
+		location.toString()
+	);
 }
 
 /**

--- a/packages/kit/types/index.d.ts
+++ b/packages/kit/types/index.d.ts
@@ -1766,7 +1766,7 @@ declare module '@sveltejs/kit' {
 	 * @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
 	 * @throws {Error} If the provided status is invalid (not between 400 and 599).
 	 */
-	export function error(status: NumericRange<400, 599>, body: App.Error): never;
+	export function error(status: number, body: App.Error): never;
 	/**
 	 * Throws an error with a HTTP status code and an optional message.
 	 * When called during request handling, this will cause SvelteKit to
@@ -1777,7 +1777,7 @@ declare module '@sveltejs/kit' {
 	 * @throws {HttpError} This error instructs SvelteKit to initiate HTTP error handling.
 	 * @throws {Error} If the provided status is invalid (not between 400 and 599).
 	 */
-	export function error(status: NumericRange<400, 599>, body?: {
+	export function error(status: number, body?: {
 		message: string;
 	} extends App.Error ? App.Error | string | undefined : never): never;
 	/**
@@ -1795,7 +1795,7 @@ declare module '@sveltejs/kit' {
 	 * @throws {Redirect} This error instructs SvelteKit to redirect to the specified location.
 	 * @throws {Error} If the provided status is invalid.
 	 * */
-	export function redirect(status: NumericRange<300, 308>, location: string | URL): never;
+	export function redirect(status: 300 | 301 | 302 | 303 | 304 | 305 | 306 | 307 | 308 | ({} & number), location: string | URL): never;
 	/**
 	 * Checks whether this is a redirect thrown by {@link redirect}.
 	 * @param e The object to check.

